pkg/dataset/dataset.go

```diff
@@ -66,71 +58,56 @@ func (d *Dataset) ListElements() []ElementMeta {
 	return elementMetas
 }
 
-func (d *Dataset) GetElement(ctx context.Context, name string) ([]byte, Element, error) {
-	e, exists := d.Elements[name]
-	if !exists {
-		return nil, Element{}, fmt.Errorf("element %s not found", name)
+func (d *Dataset) GetAllElements() []ElementNoIndex {
+	var elements []Element
+	for _, element := range d.Elements {
+		elements = append(elements, element)
 	}
-
-	contents, err := d.m.gptscriptClient.ReadFileInWorkspace(ctx, e.File, gptscript.ReadFileInWorkspaceOptions{
-		WorkspaceID: d.m.workspaceID,
+	sort.Slice(elements, func(i, j int) bool {
+		return elements[i].Index < elements[j].Index
 	})
-	if err != nil {
-		return nil, Element{}, fmt.Errorf("failed to read element %s: %w", name, err)
+
+	var noIndex []ElementNoIndex
+	for _, element := range elements {
+		noIndex = append(noIndex, ElementNoIndex{
+			ElementMeta:    element.ElementMeta,
+			Contents:       element.Contents,
+			BinaryContents: element.BinaryContents,
+		})
 	}
 
-	return contents, e, nil
+	return noIndex
 }
 
-func (d *Dataset) AddElement(ctx context.Context, name, description string, contents []byte) (Element, error) {
-	if _, exists := d.Elements[name]; exists {
-		return Element{}, fmt.Errorf("element %s already exists", name)
-	}
-
-	fileName, err := d.m.EnsureUniqueElementFilename(ctx, d.ID, toFileName(name))
-	if err != nil {
-		return Element{}, fmt.Errorf("failed to generate unique file name: %w", err)
+func (d *Dataset) GetElement(name string) (Element, error) {
+	e, exists := d.Elements[name]
+	if !exists {
+		return Element{}, fmt.Errorf("element %s not found", name)
 	}
 
-	if err := d.m.gptscriptClient.WriteFileInWorkspace(ctx, fileName, contents, gptscript.WriteFileInWorkspaceOptions{
-		WorkspaceID: d.m.workspaceID,
-	}); err != nil {
-		return Element{}, fmt.Errorf("failed to write element %s: %w", name, err)
-	}
+	return e, nil
+}
 
-	e := Element{
-		ElementMeta: ElementMeta{
-			Name:        name,
-			Description: description,
-		},
-		Index: len(d.Elements),
-		File:  fileName,
+func (d *Dataset) AddElement(e Element) error {
```

Comment:
nit: might be nice for this to be variadic; e.g. `AddElements(e ...Element) error`

Reply:
Since this is tool implementation code, people won't really interact with this, so I think I'm just going to leave it as it is. Thanks for the suggestion though